#13320: Impl uniform operation in ttnn

BuiChiTrung opened the pull request:
### Ticket
Link to Github Issue: #13320 

### Problem description
Implement pytorch Aten operation: uniform in ttnn using NPU kernels.

### What's changed
Just added code for uniform operation without changing anything. 

### How?

Expose a sfpu API. The API using hardware prng mechanism (tensix instructions) to generate a tile of random float in the range of [from, from + scale]. 

Note: there is an unavoidable problem. As reported [here](https://github.com/tenstorrent/tt-llk-wh-b0/pull/26#issuecomment-2382729826), the hardware prng is not uniform. In a tile, there are only 64 to 94 distinct numbers scattered around the tile.

### Checklist
- [x] Post commit CI passes
- [x] Blackhole Post commit (if applicable)
- [x] Model regression CI testing passes (if applicable)
- [x] Device performance regression CI testing passes (if applicable)
- [x] New/Existing tests provide coverage for changes

BuiChiTrung committed:
#13320: add draft code to test dropout op

#13320: rollback add_2_integers_in_compute changes

#13320: add draft code for uniform in programming_examples

#13320: check random numbers frequency using map

#13320: add draft code

#13320: try remove reader kernel and cb1

#13320: add skeleton code for uniform

#13220: support from, to param

#13320: remove reader kernel

#13320: add cb_intermed0

#13320: add unit-test

#13320: refactor writer kernel by exposing an sfpu API

#13320: remove native dir

#13320: refactor

#13320: update unit test

#13320: change compute_kernel_config param to const optional and add license

#13320: add cb id as compile args

#13320: raise exception when from = to

#13320: skip test for grayskull

#13320: revise uniform operation

#13320: update program factory

#13320: update unit-test for bfloat16 and refactor cb var name in kernels

#13320: add feature to benchmark ttnn with torch uniform

#13320: refactor

#13320: update compute_output_shapes func

mrshaw01 commented on ttnn/cpp/ttnn/operations/uniform/device/uniform_device_operation.cpp:
```diff
@@ -0,0 +1,64 @@
+
+UniformDeviceOperation::tensor_return_value_t UniformDeviceOperation::create_output_tensors(
+    const operation_attributes_t& operation_attributes, const tensor_args_t& tensor_args) {
+    // In-place operation, udpate input tensor directly
```
misspelling

BuiChiTrung replied:
fixed

mrshaw01 commented on ttnn/cpp/ttnn/operations/uniform/uniform.hpp:
```diff
@@ -0,0 +1,26 @@
+        const float from,
+        const float to,
+        const std::optional<ttnn::MemoryConfig> &memory_config,
+        std::optional<const ttnn::DeviceComputeKernelConfig> compute_kernel_config);
```
Please use `const std::optional<DeviceComputeKernelConfig>& compute_kernel_config` for consistency.
Also, check if can remove `ttnn::` here.

BuiChiTrung replied:
updated.

mrshaw01 commented on ttnn/cpp/ttnn/operations/uniform/uniform.cpp:
```diff
@@ -0,0 +1,21 @@
+    const float from,
+    const float to,
+    const std::optional<ttnn::MemoryConfig> &memory_config,
+    std::optional<const ttnn::DeviceComputeKernelConfig> compute_kernel_config) {
```
Please use const std::optional<DeviceComputeKernelConfig>& compute_kernel_config for consistency.

BuiChiTrung replied:
Updated

mrshaw01 commented on ttnn/cpp/ttnn/operations/uniform/uniform.cpp:
```diff
@@ -0,0 +1,21 @@
+    const float to,
+    const std::optional<ttnn::MemoryConfig> &memory_config,
+    std::optional<const ttnn::DeviceComputeKernelConfig> compute_kernel_config) {
+    return prim::uniform(input_tensor, from, to, memory_config, compute_kernel_config);
```
Must be `ttnn::prim::uniform`.

BuiChiTrung replied:
Updated

mrshaw01 commented on ttnn/cpp/ttnn/operations/uniform/device/uniform_device_operation.hpp:
```diff
@@ -0,0 +1,82 @@
+        const float from,
+        const float to,
+        const std::optional<ttnn::MemoryConfig>& memory_config,
+        std::optional<const ttnn::DeviceComputeKernelConfig> compute_kernel_config);
```
Please use const std::optional<DeviceComputeKernelConfig>& compute_kernel_config for consistency.

BuiChiTrung replied:
Updated

mrshaw01 commented on ttnn/cpp/ttnn/operations/uniform/device/uniform_device_operation.cpp:
```diff
@@ -0,0 +1,64 @@
+}
+
+void UniformDeviceOperation::validate_inputs(
+    const operation_attributes_t& attributes, const tensor_args_t& tensor_args) {
+    TT_FATAL(tensor_args.input_tensor.storage_type() == StorageType::DEVICE, "Input tensor need to be on device");
+    TT_FATAL(tensor_args.input_tensor.buffer() != nullptr, "Input tensor need to be allocated in buffers on device");
+    TT_FATAL((tensor_args.input_tensor.get_layout() == Layout::TILE), "Input tensor must be tilized");
+    TT_FATAL(
+        tensor_args.input_tensor.get_dtype() == DataType::BFLOAT16 ||
+            tensor_args.input_tensor.get_dtype() == DataType::FLOAT32,
+        "Input tensor must be Float32 or Bfloat16");
+    TT_FATAL(attributes.from <= attributes.to, "from param must be <= to");
+}
```
Please add prefix for TT_FATAL messages.
Example: "Uniform: input tensor need to be on device".

BuiChiTrung replied:
Updated

mrshaw01 commented on ttnn/cpp/ttnn/operations/uniform/device/uniform_device_operation.cpp:
```diff
@@ -0,0 +1,64 @@
+    const float from,
+    const float to,
+    const std::optional<ttnn::MemoryConfig>& memory_config,
+    std::optional<const ttnn::DeviceComputeKernelConfig> compute_kernel_config) {
```
const std::optional& compute_kernel_config

BuiChiTrung replied:
Updated

hschoi4448 commented on tt_metal/hw/ckernels/wormhole_b0/metal/llk_api/llk_sfpu/ckernel_sfpu_rand_uint.h:
```diff
@@ -0,0 +1,29 @@
+#pragma once
```
```suggestion
// SPDX-FileCopyrightText: © 2024 Tenstorrent Inc.
//
// SPDX-License-Identifier: Apache-2.0
#pragma once
```

BuiChiTrung replied:
Added.

hschoi4448 commented on ttnn/cpp/ttnn/operations/uniform/device/kernels/writer.cpp:
```diff
@@ -0,0 +1,73 @@
+        cb_wait_front(cb_intermed0_id, 1);
+
+        uint32_t cb_intermed0_read_ptr = get_read_ptr(cb_intermed0_id);
+        uint32_t cb_out0_write_ptr = get_write_ptr(cb_out0_id);
```
```C++
 * Returns a pointer to the beginning of a memory block previously reserved
 * by cb_reserve_back. Note that this call is only valid between calls
 * to cb_reserve_back and cb_push_back. The amount of valid memory
 * is equal to the number of tiles requested in a prior cb_reserve_back call.
 uint32_t get_write_ptr(uint32_t operand) {
```
https://github.com/tenstorrent/tt-metal/blob/main/tt_metal/hw/inc/dataflow_api.h#L399

According to the comment, `get_write_ptr` is only valid between `cb_reserve_back` and `cb_push_back`.
Even if there's no actual problem, it would be better to place cb_reserve_back and cb_push_back somewhere in the code for clarity and safety.

BuiChiTrung replied:
I've added `cb_reserve_back`, but `cb_push_back` is not added as this cb is only used in 1 core for write purpose. If I add `cb_push_back` the program will hang as there is no other core executing `cb_wait_front`. The same approach can be founded in this file: ttnn/cpp/ttnn/operations/data_movement/fill_rm/device/kernels/dataflow/fill_rm_interleaved.cpp

hschoi4448 replied:
I understand that we only need to add `cb_reserve_back`. However, I don't see `cb_reserve_back` in the code. It seems that the commit might be missing it.

BuiChiTrung replied:
`cb_reserve_back` is added in this [commit](https://github.com/tenstorrent/tt-metal/pull/13735/commits/f07514be68e8d056e929063eebf882750f50e812#diff-094c033c84cbdc24390a2b6b167f2531cd87ffb3309689586f37916fc0fdbdb0), line 37

hschoi4448 replied:
Got it. It seems I might have been looking at an outdated file.

hschoi4448 commented on ttnn/cpp/ttnn/operations/uniform/device/kernels/writer.cpp:
```diff
@@ -0,0 +1,73 @@
+                // The hardware PRNG is not uniformly distribute.
+                // Generated rand_floats in range (0, 0.5] has higher ratio compared to (0.5, 1).
+                // I *2 rand_float < 0.5 to make it more uniform.
+                if (rand_float < 0.5) {
```
```suggestion
                if (rand_float < 0.5f) {
```
In C++, `0.5` is a `double` by default, so using `0.5f` is better when you need a `float` type for precision and performance reasons.

BuiChiTrung replied:
Updated.

hschoi4448 commented on ttnn/cpp/ttnn/operations/uniform/device/kernels/writer.cpp:
```diff
@@ -0,0 +1,73 @@
+                uint32_t rand_uint32 = *cb_intermed0_addr;
+                float rand_float = static_cast<float>(rand_uint32) / max_uint;
+                // The hardware PRNG is not uniformly distribute.
+                // Generated rand_floats in range (0, 0.5] has higher ratio compared to (0.5, 1).
```
```suggestion
                // Generated rand_floats in range [0, 0.5] has higher ratio compared to (0.5, 1].
```
Based on the existing explanation of the ranges `(0, 0.5]` and `(0.5, 1)`, it seems that 0 and 1 are excluded from the output range.
If 0 and 1 need to be included, modifying the ranges as suggested should address the issue.

hschoi4448 replied:
Looking at the implementation of PyTorch's `rand`, it includes the minimum and excludes the maximum. Please check if 1 is included in the uniform output.

BuiChiTrung replied:
<img width="869" alt="image" src="https://github.com/user-attachments/assets/6f61b3b0-d11f-4d6e-9d29-1b7c39da43f0">
I tried on pytorch, with from=to=5. The output tensor contains only 5. So I think both from to can be included in the result.

hschoi4448 replied:
![image](https://github.com/user-attachments/assets/d612e624-890a-47fa-9de0-ab3405fbba26)
![image](https://github.com/user-attachments/assets/409bfe0b-4683-4ffd-ba16-424562285a3f)
According to the documentation, it seems that "exclude" is correct. The case where low equals high appears to be a special case.

BuiChiTrung replied:
These are 2 seperate operations. I try `torch.distributions.uniform` with low = high = 5.0 and it raise an error.

<img width="928" alt="image" src="https://github.com/user-attachments/assets/638a8b65-7ef5-4e06-a981-61476d56edf5">

Which implementation should I follow `torch.Tensor.uniform_` or `torch.distributions.uniform`

BuiChiTrung replied:
I think torch.Tensor.uniform_ is more reasonable as other operations in the epic are in torch or torch.Tensor module.

hschoi4448 replied:
```Python
import torch

torch.manual_seed(0)

# Tensor.uniform_ example
tensor1 = torch.empty(3, 3)
tensor1.uniform_(0, 1)

torch.manual_seed(0)
# distributions.uniform example
distribution = torch.distributions.uniform.Uniform(0, 1)
samples = distribution.sample((3, 3))

print("Tensor.uniform_ result:")
print(tensor1)
print("distributions.uniform result:")
print(samples)

----------------
Tensor.uniform_ result:
tensor([[0.4963, 0.7682, 0.0885],
        [0.1320, 0.3074, 0.6341],
        [0.4901, 0.8964, 0.4556]])
distributions.uniform result:
tensor([[0.4963, 0.7682, 0.0885],
        [0.1320, 0.3074, 0.6341],
        [0.4901, 0.8964, 0.4556]])
```
It seems that `Tensor.uniform_` and `distributions.uniform.Uniform` are the same function. 
The reason for the Python `ValueError` might be that the distributions version has one more validation step.

hschoi4448 replied:
![image](https://github.com/user-attachments/assets/5d300512-afd7-4eb4-8a08-fc5e9cf3f0ed)

https://github.com/pytorch/pytorch/blob/main/torch/distributions/uniform.py#L79
The sample method of distributions.uniform.Uniform calls rsample on line 79. Looking at rsample, it calls `torch.rand`, which is also defined in the range [0, 1).

BuiChiTrung replied:
I'm still a little bit confused but since this is the behavior of torch.rand(), I will follow. Updated in this [commit](https://github.com/tenstorrent/tt-metal/pull/13735/commits/331cd75c78cba9765c567c53a9fcdd9921b0e9d0).

hschoi4448 commented on ttnn/cpp/ttnn/operations/uniform/device/uniform_program_factory.cpp:
```diff
@@ -0,0 +1,153 @@
+    Program program = Program();
+
+    DataType output_dtype = output.dtype();
+    tt::DataFormat out_data_format = datatype_to_dataformat_converter(output_dtype);
```
```suggestion
    DataFormat out_data_format = datatype_to_dataformat_converter(output_dtype);
```

BuiChiTrung replied:
<img width="587" alt="image" src="https://github.com/user-attachments/assets/a44dddb9-ec30-4031-9be5-90b7a66b2cab">
There are 2 candidates for DataFormat. I have to use tt:: to specify which one to use.

hschoi4448 replied:
Wow, this is another issue. For now, I think we can change it to "auto."

BuiChiTrung replied:
Updated

BuiChiTrung commented on ttnn/cpp/ttnn/operations/uniform/device/uniform_device_operation.cpp:
```diff
@@ -0,0 +1,64 @@
+// SPDX-FileCopyrightText: © 2024 Tenstorrent Inc.
```
Updated

BuiChiTrung commented on ttnn/cpp/ttnn/operations/uniform/device/uniform_device_operation.cpp:
```diff
@@ -0,0 +1,64 @@
+            to,
+            memory_config.value_or(input_tensor.memory_config()),
+            init_device_compute_kernel_config(
+                input_tensor.device()->arch(), compute_kernel_config, MathFidelity::HiFi4, true, true)},
```
I forgot mentioning this in the PR desc. In this op, default_fp32_acc must be set to true. Otherwise, generated random float would always be in the range (0.4, 0.5).

hschoi4448 commented on tests/ttnn/unit_tests/operations/test_uniform.py:
```diff
@@ -0,0 +1,109 @@
+    logger.info(f"Expected mean: {expected_mean}, NPU mean: {npu_mean}")
+    logger.info(f"Expected var: {expected_var}, NPU var: {npu_var}")
+
+    assert rand_from <= min_val and max_val <= rand_to
```
```suggestion
    assert rand_from <= min_val and max_val < rand_to
```
I understand that uniform does not include the rand_to value.

BuiChiTrung replied:
Same as this cmt: https://github.com/tenstorrent/tt-metal/pull/13735#discussion_r1796747564

BuiChiTrung replied:
Updated in this [commit](https://github.com/tenstorrent/tt-metal/pull/13735/commits/331cd75c78cba9765c567c53a9fcdd9921b0e9d0).

mrshaw01 commented:
I have already created a revised version of your PR.  
Check it out here: [mrshaw01/revise_uniform_operation](https://github.com/tenstorrent/tt-metal/tree/mrshaw01/revise_uniform_operation)

1. I've revised all the minor details and also changed the naming for all kernels and the program factory for consistency. You can view the differences in the latest commits.

2.  
You did a great job handling `OUTPUT_DTYPE_BFLOAT16` in `write_uniform.cpp`. From what I saw, you're truncating the 16 mantissa bits. However, the issue arises when truncating the mantissa bits, which leads to a rounding problem. As a result, the value can end up lower than `low` or higher than `high`.

3.  
I added a TODO in `uniform_program_factory.cpp`: `// TODO(Shaw): Eliminate the need for this trigger.` I tested changing `CB::c_out0` to `CB::c_in0`, and it still works.

4.  
I don’t fully understand this part yet:
```C
    // Init a nice per-lane counter
    TTI_SFPMOV(0, p_sfpu::LTILEID, p_sfpu::LREG0, 0);
    TTI_SFPSHFT(-1 & 0xfff, 0, p_sfpu::LREG0, 1);
```
Let's discuss it at the office tomorrow.

Please review the revised version, and we can discuss it further tomorrow.

cc. @hschoi4448

BuiChiTrung commented:
> I have already created a revised version of your PR. Check it out here: [mrshaw01/revise_uniform_operation](https://github.com/tenstorrent/tt-metal/tree/mrshaw01/revise_uniform_operation)
> 
> 1. I've revised all the minor details and also changed the naming for all kernels and the program factory for consistency. You can view the differences in the latest commits.
> 
> You did a great job handling `OUTPUT_DTYPE_BFLOAT16` in `writer_uniform.cpp`. From what I saw, you're truncating the 16 mantissa bits. However, the issue arises when truncating the mantissa bits, which leads to a rounding problem. As a result, the value can end up lower than `low` or higher than `high`.
> 
> 
> I added a TODO in `uniform_program_factory.cpp`: `// TODO(Shaw): Eliminate the need for this trigger.` I tested changing `CB::c_out0` to `CB::c_in0`, and it still works.
> 
> 4. The kernels were revised so that the APIs are used properly. Please check them carefully.
> 
> I don’t fully understand this part yet:
> 
> ```c
>     // Init a nice per-lane counter
>     TTI_SFPMOV(0, p_sfpu::LTILEID, p_sfpu::LREG0, 0);
>     TTI_SFPSHFT(-1 & 0xfff, 0, p_sfpu::LREG0, 1);
> ```
> 
> Let's discuss it at the office tomorrow.
> 
> Please review the revised version, and we can discuss it further tomorrow.
> 
> cc. @hschoi4448

I've merged 2 branches.
1. Checked.
2.  I just update the unit-test threshold to avoid asserting fail. That is un-avoidable error, even torch can't: check this example with bfloat16, some elements are < from param.

<img width="863" alt="image" src="https://github.com/user-attachments/assets/070fe63c-e83a-4754-a572-0b4155fed636">

3. You're right. Updated to use only cb_in0.
4. Checked.
5. Tbh, I copied that part from another example to generate random uint32 [here](https://github.com/tenstorrent/tt-llk-wh-b0/pull/26#issuecomment-2382729826).

BuiChiTrung commented:
> I have already created a revised version of your PR. Check it out here: [mrshaw01/revise_uniform_operation](https://github.com/tenstorrent/tt-metal/tree/mrshaw01/revise_uniform_operation)
> 
> 1. I've revised all the minor details and also changed the naming for all kernels and the program factory for consistency. You can view the differences in the latest commits.
> 
> You did a great job handling `OUTPUT_DTYPE_BFLOAT16` in `writer_uniform.cpp`. From what I saw, you're truncating the 16 mantissa bits. However, the issue arises when truncating the mantissa bits, which leads to a rounding problem. As a result, the value can end up lower than `low` or higher than `high`.
> 
> 
> I added a TODO in `uniform_program_factory.cpp`: `// TODO(Shaw): Eliminate the need for this trigger.` I tested changing `CB::c_out0` to `CB::c_in0`, and it still works.
> 
> 4. The kernels were revised so that the APIs are used properly. Please check them carefully.
> 
> Please review the revised version, and we can discuss it further tomorrow.
> 
> cc. @hschoi4448

I've merged 2 branches.
1. Checked.
2.  I just update the unit-test threshold to avoid asserting fail. That is un-avoidable error, even torch can't: check this example with bfloat16, some elements are < from param.

<img width="863" alt="image" src="https://github.com/user-attachments/assets/070fe63c-e83a-4754-a572-0b4155fed636">

3. You're right. Updated to use only cb_in0.
4. Checked.

hschoi4448 commented on tests/ttnn/unit_tests/operations/test_uniform.py:
```diff
@@ -0,0 +1,127 @@
+    assert np.allclose(npu_var, expected_var, rtol=0.5)
+
+
+# fmt: off
```
```suggestion
```
Is this really necessary?

BuiChiTrung replied:
It used so that the IDE doesn't auto format on save that block of code which merges my test param into one line. I will remove it before merging the branch. However, for now, it's convenient as in the developing/reviewing phase, I constantly need to comment/un-comment a specific test.

BuiChiTrung commented:
@razorback3 Please take a look if there is anything need to be updated.

ayerofieiev-tt reviewed (approved):
@rdjogoTT can you please also give this a look? Thank you!

BuiChiTrung committed:
#13320: update unit-test

BuiChiTrung committed:
#13320: update sfpu api

rtawfik01 commented on tt_metal/hw/ckernels/wormhole_b0/metal/llk_api/llk_sfpu/ckernel_sfpu_rand_uint.h:
```diff
@@ -0,0 +1,32 @@
+
+#include "ckernel.h"
+#include "ckernel_defs.h"
+#include "ckernel_sfpu_converter.h"
```
Can this `"ckernel_sfpu_converter.h"` be removed since it is not used in this file?

rtawfik01 commented on tt_metal/hw/ckernels/wormhole_b0/metal/llk_api/llk_sfpu/ckernel_sfpu_rand_uint.h:
```diff
@@ -0,0 +1,32 @@
+// SPDX-FileCopyrightText: © 2024 Tenstorrent Inc.
```
Can this also be added for Blackhole so we have same set of operations for both devices?

BuiChiTrung replied:
Frankly, I don't have access to the Blackhole device. I can't test if this would also work on Blackhole.

rtawfik01 replied:
ok, can you make an issue that is missing impl for BH after you merge this PR? And make sure to add the "blackhole" tag, and to also add the project "Blackhole"

BuiChiTrung replied:
Ok, got it.

rtawfik01 commented on tt_metal/hw/ckernels/wormhole_b0/metal/llk_api/llk_sfpu/ckernel_sfpu_rand_uint.h:
```diff
@@ -0,0 +1,32 @@
+        TTI_SFPMOV(0, 9, p_sfpu::LREG0, 8);
+        TTI_SFPSETSGN(0, p_sfpu::LREG0, p_sfpu::LREG0, 1);
+        TTI_SFPSTORE(0, 4, 3, 0);
+        TTI_SFPIADD(0, p_sfpu::LREG0, p_sfpu::LREG0, 5);
```
@BuiChiTrung Can you explain in the above how the `SFPSTORE` function is used before `SFPIADD` and it is giving the correct results? The `SFPSTORE` function is the one that writes the lreg (in the case above it is LREG0) to dst. So I am confused on why the SFPIADD is called after the result is written back to dst?

rtawfik01 replied:
or is the LREG0 above maintaining value like a counter?

BuiChiTrung replied:
You're right. `LREG0` is used as a counter, I removed the counter init code in this [commit](https://github.com/tenstorrent/tt-metal/pull/13735/commits/6837a87e6b0fdf175f7994dcaf5741805f494b99). It's my bad not removing `SFPIADD` here. Just removed it.

BuiChiTrung committed:
#13320: remove un-used header in kernel and counter in sfpu API

rdjogoTT commented on tt_metal/hw/ckernels/wormhole_b0/metal/llk_api/llk_sfpu/ckernel_sfpu_rand_uint.h:
```diff
@@ -0,0 +1,30 @@
+#pragma GCC unroll 0
+    for (int d = 0; d < 8; d++) {
+        TTI_SFPMOV(0, 9, p_sfpu::LREG0, 8);
+        TTI_SFPSETSGN(0, p_sfpu::LREG0, p_sfpu::LREG0, 1);
```
Can you explain why the SFPSETSGN instruction is needed? Is this related to your comments in the writer kernel "The hardware PRNG is not uniformly distribute."

rdjogoTT replied:
The sign bit is just the MSB, and when working with uint32 numbers it isn't used as a sign bit, it's actually used as 2^31. So by zeroing it you would limit the range to be half of uint32 max value.

rtawfik01 replied:
@BuiChiTrung without the `TTI_SFPSETSGN` instruction, were you seeing negative output from the prng? Or did you use `TTI_SFPSETSGN` as just in case?

BuiChiTrung replied:
Just checked again. I saw negative numbers in first place because I mistakenly used Int32 CB to get result from sfpu. Updated [here](https://github.com/tenstorrent/tt-metal/pull/13735/commits/992a2c0c2af023fd81e1a82b35c8ae4a1258e70c).

BuiChiTrung committed:
#13320: remove TTI_SFPSETSGN instr

ttmtrajkovic commented:
@BuiChiTrung,
I'd like @amahmudTT to review this code as he worked on internal implementation of the uniform / dropout op.

@amahmudTT,
could you please take a look at the implementation details?

Milos

BuiChiTrung committed:
#13320: allow set fp32_dest_acc to false

BuiChiTrung committed:
#13320: update callback test

BuiChiTrung committed:
#13320: skip test for grayskull

amahmudTT commented on ttnn/cpp/ttnn/operations/uniform/device/uniform_program_factory.cpp:
```diff
@@ -0,0 +1,147 @@
+using namespace tt::tt_metal;
+
+std::mt19937 rng(std::time(nullptr));
+std::uniform_int_distribution distribution(1, 1 << 20);
```
why 1 << 20 and not the full range of uint ?

BuiChiTrung replied:
Initially, I used `rand_uint_tile_init(i * seed)` with `i` is the tile_id. `1 << 20` used to avoid seed overflow. However, as I've updated the code to init only once, `1 << 20` is updated to max_int (`distribution` required 2nd param to be int).

amahmudTT commented on ttnn/cpp/ttnn/operations/uniform/device/kernels/writer_uniform.cpp:
```diff
@@ -0,0 +1,68 @@
+    uint32_t num_tiles = get_arg_val<uint32_t>(4);
+    uint32_t end_id = start_id + num_tiles;
+
+    const InterleavedAddrGenFast<output_is_dram> output_addrg = {
```
I think this could be done on the SFPU and it would be much faster.  What if we generate a random number on SFPU then shift it 9 bits to the right, so that  we get the mantissa part of a fp32 number, then in the exponent section we put the appropriate exponent such that our number is between 1 and 2, and then we subtract 1 from it using a programmable constant register. Then we have a value between 0 and 1 . 

Then we use eltwise operations to multiply with the range and then add to the from constant. This may require careful consideration of the bounds, but this will be much faster as things will be done in parallel and without trips to L1

BuiChiTrung replied:
Thank for your suggestion, I've updated the sfpu API to generate random float instead of uint32.

amahmudTT commented on ttnn/cpp/ttnn/operations/uniform/device/kernels/writer_uniform.cpp:
```diff
@@ -0,0 +1,68 @@
+        for (uint32_t k = 0; k < constants::TILE_WIDTH; k++) {
+            for (uint32_t j = 0; j < constants::TILE_HEIGHT; j++) {
+                uint32_t rand_uint32 = *intermed_cb_addr;
+                float rand_float = static_cast<float>(rand_uint32) / max_uint;
```
Is it possible to control the rounding here to make sure it is round towards zero and not round towards nearest as that may exceed the range ?

BuiChiTrung replied:
This code is removed as I've updated sfpu API to generate random float.

amahmudTT reviewed (commented):
I think the whole process could be done in the compute thread, that is instead of just parallelizing the generation of random numbers we could also parallelize the uniform number creation. 

The steps would be 

- Generate random number 
- Shift random number by 9 bits (to treat it as float with 0 sign bit and empty exponent for now) 
- Put 127 in the exponent field using the SETEXP instruction, now our value as a float is between 1 and 2 (actually 2 - 2^-24 but you can probably ignore this error since it is also present when an integer is cast to float in the current code. Minor optimization here could allow for skipping the shift) 
- Subtract 1 from the result.  
- Multiply with range. 
- Add From

BuiChiTrung committed:
#13320: update sfpu to generate random float

BuiChiTrung committed:
#13320: update sfpu to generate random float between range [from, to)

amahmudTT commented on tt_metal/hw/ckernels/wormhole_b0/metal/llk_api/llk_sfpu/ckernel_sfpu_rand.h:
```diff
@@ -16,20 +16,24 @@ inline void rand_init(uint seed) {
+
+        vFloat v_from = from;
+        vFloat v_to = to;
+        rand_floats = (rand_floats - vConst1) * (v_to - v_from) + v_from;
```
You could calculate v_to - v_from outside the loop and save it in a register and avoid the recalculation every time inside the loop.  May be you could put it into one of those programmable constants in the init phase, instead of passing them as arguments.

BuiChiTrung replied:
It seems like that programmable constants can only be created in the face loop.

<img width="1219" alt="image" src="https://github.com/user-attachments/assets/b94dc4e4-65b4-44b7-bf53-5823527cb061">

amahmudTT replied:
It needs to be in the initialization I think, see file ckernel_sfpu_recip.h

```
template <bool APPROXIMATION_MODE>
inline void _init_reciprocal_()
{
    vConstFloatPrgm0 = 1.442695f; // ln2_recip
    vConstFloatPrgm1 = 2.0f;
}

} // namespace sfpu
} // namesp
```

amahmudTT replied:
If you want to avoid doing it in the initialization then I think you need to pass the float value's bit patterns as uint32_t and then read them outside the loop and calculate the range, see ckernel_sfpu_dropout.h 

```
// probability should be between 0 - INT_MAX (signed)
// scale should be binary representation of a float32
template <bool APPROXIMATION_MODE, int ITERATIONS>
inline void _calculate_dropout_(const int iterations, uint probability, uint scale)
{
    // SFPU microcode

    FWLOG1("calculate_dropout() -- probability:%x", probability);
    FWLOG1("calculate_dropout() -- scale:%x", scale);

    TT_SFPLOADI(p_sfpu::LREG1, 10, scale&0xFFFF);
    TT_SFPLOADI(p_sfpu::LREG1, 8, scale>>16);
    TT_SFPLOADI(p_sfpu::LREG2, 10, probability&0xFFFF);
    TT_SFPLOADI(p_sfpu::LREG2, 8, probability>>16);
    #pragma GCC unroll 0
    for (int d = 0; d < iterations; d++) {

        ////////////////////////
        ```
        
        Here I am doing some pre computation outside the loop.

amahmudTT replied:
Given I am not using immediate, I cannot use TTI_SFPLOADI, I have to use TT_SFPLOADI which is slower since trisc core cannot just pass it to the tensix core, it has to create the instruction. The instructions can be directly passed from the trisc core to tensix core if they all use immediates

BuiChiTrung replied:
Thank for your explanation, I've update the ckernel file accordingly.

amahmudTT commented on tt_metal/hw/ckernels/wormhole_b0/metal/llk_api/llk_sfpu/ckernel_sfpu_rand.h:
```diff
@@ -16,20 +16,24 @@ inline void rand_init(uint seed) {
 
 template <bool APPROXIMATION_MODE>
-inline void rand() {
+inline void rand(float from, float to) {
```
nit, randf should sound more correct than rand.

BuiChiTrung replied:
I follow torch, c++ API, all of which use rand() to generate random float.

BuiChiTrung committed:
#13320: refactor writer kernel

BuiChiTrung commented on tests/ttnn/unit_tests/operations/test_uniform.py:
```diff
@@ -0,0 +1,160 @@
+
+    """
+    Random bfloat16 is converted from random float. As 16 bits are truncated, the generated number might be smaller/bigger than from/to.
+    (even torch can't handle that case, check check_torch_uniform_bfloat16() function). I use bfloat16_ep is used to avoid asserting fail.
```
Updated

BuiChiTrung replied:
This var is no longer used. I've changed the assert method by comparing the generated number with rand_from, rand_to (but casted to bfloat16/float32)

BuiChiTrung commented on tt_metal/hw/ckernels/wormhole_b0/metal/llk_api/llk_sfpu/llk_math_eltwise_unary_sfpu_rand_uint.h:
```diff
@@ -0,0 +1,26 @@
+// SPDX-FileCopyrightText: © 2023 Tenstorrent Inc.
```
Updated

BuiChiTrung commented on ttnn/cpp/ttnn/operations/uniform/device/kernels/compute_uniform.cpp:
```diff
@@ -0,0 +1,37 @@
+    init_sfpu(intermed_cb_id);
+
+    for (uint32_t i = start_id; i < end_id; ++i) {
+        rand_uint_tile_init(i * seed);
```
You're right. Updated.

BuiChiTrung committed:
#13320: update compute kernel

amahmudTT commented on tt_metal/hw/ckernels/wormhole_b0/metal/llk_api/llk_sfpu/ckernel_sfpu_rand.h:
```diff
@@ -0,0 +1,42 @@
+        TTI_SFPSTORE(0, 4, 3, 0);
+
+        // -1 to ensure the float is within the range [0, 1).
+        vFloat rand_floats = dst_reg[0];
```
In general it is recommended not to mix SFPI and the TTI_* instructions and to write them using only one form. 
Do you need to save to dst and reload from it ? Is it not possible to have all the values in the registers and do one final write to dst ? That would save you some cycles I think.

BuiChiTrung replied:
That's what I want to ask you. I tried arithmetic instruction like: `TTI_SFPADDI(1, p_sfpu::LREG0, 0);` before `TTI_SFPSTORE` to -1 to rand_float but the generated numbers are not changed. That's why I have to update use `vFloat`

This is the first time I deal with tensix instruction. Is there anything I'm doing wrong here with the arithmetic instructions.

amahmudTT replied:
The thing is the compiler will not do anything with the TTI_* instructions and keep them as they are, but it may try to optimize the vFloat instructions, so there might be register reordering or something, but in your case you are writing to dst and then reading from dst, so there is not much for the compiler to move around or change so should be okay. But the thing is when using TTI instructions you know exactly which registers you are using so you may end up fitting all the data within the registers and not need to read and write from the dst for intermediate values saving some cycles.

amahmudTT replied:
@BuiChiTrung  SFPADDI requires 2 cycles to complete, so you need to put a NOP after it  otherwise the result will be incorrect if an instruction uses it right after.

BuiChiTrung replied:
I tried with NOP(), the value is updated. I tried pass various of value as `imm16_math` of SFPADDI param but it only return some 32x32 weird value of -nan.
```
TTI_SFPADDI(-1, p_sfpu::LREG0, 0);
TTI_SFPNOP;
```
If it's possible, could you suggest the code to convert those statements with vFloat to TTI_*. As there is no document with TTI_, it's really hard converting sfpi to TTI_, which requires a lot of trial and error.

amahmudTT replied:
You need to put the bit pattern of -1 in bfp16 format. So you can save -1 in bfp16 format, print the unsigned int value for that bit pattern and put that value here.

BuiChiTrung replied:
Got that. I've update the ckernel.

amahmudTT commented on ttnn/cpp/ttnn/operations/uniform/device/kernels/writer_uniform.cpp:
```diff
@@ -0,0 +1,59 @@
+#endif
+
+#ifdef OUTPUT_DTYPE_BFLOAT16
+        auto dst_cb_addr = reinterpret_cast<uint8_t *>(dst_cb_write_ptr);
```
Is this step not possible on the SFPU I mean conversion from fp32 to bfp16  using SFPSTORE instruction mod 2 ?

BuiChiTrung replied:
I can't as I have to use vFloat after TTI_SFPSTORE (due to the issue described [in this cmt](https://github.com/tenstorrent/tt-metal/pull/13735#discussion_r1810535669)).

BuiChiTrung replied:
I tried mod 2, the output tile look likes this: 512 numbers in range (0, 1), 512 zeros. Do you have any idea why?

<img width="1836" alt="image" src="https://github.com/user-attachments/assets/3193ca58-dfa4-416e-9113-dab0d43459e7">

Also, I found no ckernel example use mode 2. Found no ckernel file handle 2 dtypes, neither.

amahmudTT replied:
Do you change the type of your circular buffer depending on whether you are writing fp32 of bfp16 ?

BuiChiTrung replied:
Yes, I do. Please check this [draft commit](https://github.com/tenstorrent/tt-metal/pull/13735/commits/71a1528d602ce45c5872ffcb67f7c8ba8400ae4c).

amahmudTT replied:
Did it work in your draft commit ?

BuiChiTrung replied:
No, it didn't. Half of the output are zero as the image above.

amahmudTT replied:
If you use fp32 and use the other TTI_SFPSTORE  with mod 0 or 3 do you get the full tile ?

BuiChiTrung replied:
Here is what I've tried
- mod = 2: get half bfloat16 tile.
- mod = 4: get full float32 tile.
- mod = 0: get full float32, bfloat16 tile. However, there is a precision error with bfloat16 which may cause the generated number to be equal to `to` param.

amahmudTT replied:
Okay for bfloat16 can you use mod 0 but add another instruction to zero out the last 16 bits and see if you see the error. I will look a bit more into it to see how we can properly use mod=2.

amahmudTT replied:
If you are not out of registers you can 
TT_SFPLOADI(p_sfpu::LREG1, 8, 0XFFFF);
and then do the and  with the register containing the high 16 bits.

amahmudTT replied:
or     
    TT_SFPLOADI(p_sfpu::LREG1, 10, 0);
    TT_SFPLOADI(p_sfpu::LREG1, 8, 0XFFFF);

BuiChiTrung replied:
I tried zero-out last 16 bits, the precision err remains, still. 

Another issue with mod 0 or 2, with bfloat16 is that there are more chances random number are duplicate (which idk why). In an input tensor of [1024, 1024] shape, there are only 268 distinct elements whereas with the initial approach of casting to bfloat16 in writer kernel we could reach more than 30000 distinct elements.

If there is no update, I think I'll stick with the initial approach of casting to bfloat16 in the writer kernel.

amahmudTT replied:
Ok that sounds good.

amahmudTT replied:
The SFPU rounds to the nearest even, so some rounding issue is expected. But I was hoping zeroing out the last 16 bits would help.

amahmudTT replied:
But I meant zeroing out the last 16 bits of the random number between 0-1 though. Not the final result.

amahmudTT replied:
More duplicate random numbers, I think makes sense, as we are taking only a small fraction of the random bit pattern , so there are more duplicate cases, but when you use the full range in fp32 that is 23 bits of randomness, after multiplication and addition you have more uniqueness. So I guess your approach would fare well in terms of randomness.

BuiChiTrung replied:
I've rollbacked the code to the initial approach. Please take another look. Thank you.

amahmudTT replied:
Thanks a lot @BuiChiTrung  I will take a look , but could you tell me how I can reproduce the error you mentioned when you used the SFPSTORE for fp16 ? We might want to find out a workaround and such a test might help us not ignore the error. Thanks.

BuiChiTrung replied:
You can checkout this [commit](https://github.com/tenstorrent/tt-metal/pull/13735/commits/e2d9767ad79c03818b21dcbc48fbb6db934783e8), turn on the DPRINT, run the unit-test to see a tile of bfp16 is printed with 512 zeros.

amahmudTT replied:
I meant the error that it reaches the parameter ```to```,  is it triggered by the existing test cases you added ?

BuiChiTrung replied:
Oh, for that bug. You can change SFPSTORE instr_mod to 0, use the same test but with bigger shape [100, 100].

So can I merge the code using the current approach? If so, please approve this PR.

amahmudTT replied:
@BuiChiTrung when you said when you used SFPSTORE mod 0 and bfp16, you reach the TO parameter. Did you reach it when the range was all negative ? like [-200, -100) or did you reach it with a positive TO. If you reached it with a positive TO then we need to investigate it since it is not supposed to happen.

amahmudTT replied:
And did you create the bug ? Could you share the issue number ?

BuiChiTrung replied:
TO is reached when it's positive. I haven't tried with negative range.

BuiChiTrung replied:
FYI, the sfpu API create random number in range `[from, from + scale]`. To make sure that generated number < `to` (from +scale), in the uniform program factory, I minus to with an epsilon: `-1e6`. 
```
const float eps = 1e-6;
        union {
            float f;
            uint32_t u;
        } f2u_from, f2u_to;
        f2u_from.f = operation_attributes.from;
        f2u_to.f = operation_attributes.to - eps;  // -eps make sure that generated number is < operation_attributes.to
```

But with mod 0, `to` is reached.

BuiChiTrung committed:
#13320: update sfpu to use TTI_ only

BuiChiTrung committed:
#13320: change cb format to float32

BuiChiTrung commented on tests/ttnn/unit_tests/operations/test_uniform.py:
```diff
@@ -0,0 +1,160 @@
+    logger.info(f"Expected mean: {expected_mean}, NPU mean: {npu_mean}")
+    logger.info(f"Expected var: {expected_var}, NPU var: {npu_var}")
+
+    bfloat16_ep = 0.015
```
Same as the cmt above.

BuiChiTrung commented on tests/ttnn/unit_tests/operations/test_uniform.py:
```diff
@@ -0,0 +1,160 @@
+@pytest.mark.parametrize("rand_range",
+    [
+        [2.1, 9],
+        [-5.1, 1.2] # negative float range
```
If eps are small enough, as I've tested with eps = 0.05, rand float32 is fine but rand bfloat16 will be out of bound and there is no way fix that.

BuiChiTrung committed:
#13320: update var name

amahmudTT commented on tests/ttnn/unit_tests/operations/test_uniform.py:
```diff
@@ -0,0 +1,154 @@
+    logger.info(input)
+
+
+# ttnn is faster than torch if the tensor is big enough: [1024, 1024]
```
This comment will need to be updated when the PR is done, I think the final version will be faster.

BuiChiTrung replied:
I've updated the comment.

BuiChiTrung committed:
#13320: draft commit to support generate random bfloat16 in sfpu

amahmudTT commented on tt_metal/hw/ckernels/wormhole_b0/metal/llk_api/llk_sfpu/ckernel_sfpu_rand.h:
```diff
@@ -0,0 +1,51 @@
+        TTI_SFPMAD(p_sfpu::LREG0, p_sfpu::LREG1, p_sfpu::LREG2, p_sfpu::LREG0, 1);
+        TTI_SFPNOP;
+
+        TTI_SFPSTORE(0, 2, 3, 0);
```
I think you have to use a define or preprocessor here to control there SFPSTORE,   the instruction mod 2 should be used when you want to store fp16b , otherwise 3 can be used to indicate fp32, 0 will decide based on the input data type.

amahmudTT replied:
#ifdef OUTPUT_DTYPE_FLOAT32
                   TTI_SFPSTORE(0, 3, 3, 0); / TTI_SFPSTORE(0, 0, 3, 0);
        #endif
                    TTI_SFPSTORE(0, 2, 3, 0);
        #ifdef OUTPUT_DTYPE_BFLOAT16

BuiChiTrung committed:
#13320: rollback to inital approach to handle bfloat16

amahmudTT commented:
Approved. When you merge please create an issue for optimizing the bfp16 part while making sure the bugs are not hit. 
Thanks a lot for patiently addressing the review comments and trying out different code paths !

BuiChiTrung commented:
> Approved. When you merge please create an issue for optimizing the bfp16 part while making sure the bugs are not hit (repeated random number, reaching the `to` in the `[from,to)` range) Thanks a lot for patiently addressing the review comments and trying out different code paths !

I'll create the issue. Thank you for the detailed review and the suggestions on optimizing the code. This has greatly improved my understanding of TTI.

BuiChiTrung committed:
#13320: rebase with main

BuiChiTrung committed:
#13320: rm unused header
